From ce259c2ec58843c26e63f86d7eae053029168886 Mon Sep 17 00:00:00 2001
From: Andy Doan <andy@foundries.io>
Date: Fri, 13 Sep 2019 16:06:49 -0500
Subject: [PATCH 02/11] dockerapp: Check verifyTarget no matter what

The Target given to the dockerapp package manager is always an Ostree
target. That target's custom data are all the apps we need to verify.
This makes sure we check those if needed and includes a simple change
to CI to help make sure we can handle those failures.

Signed-off-by: Andy Doan <andy@foundries.io>
---
 src/libaktualizr/package_manager/docker_fake.sh           | 7 ++++++-
 src/libaktualizr/package_manager/dockerappmanager.cc      | 7 +++++--
 src/libaktualizr/package_manager/dockerappmanager_test.cc | 3 +++
 3 files changed, 14 insertions(+), 3 deletions(-)

diff --git a/src/libaktualizr/package_manager/docker_fake.sh b/src/libaktualizr/package_manager/docker_fake.sh
index 4e50b6ff..31c6d32f 100755
--- a/src/libaktualizr/package_manager/docker_fake.sh
+++ b/src/libaktualizr/package_manager/docker_fake.sh
@@ -1,5 +1,10 @@
 #! /bin/bash
-set -eEuo pipefail
+set -eEo pipefail
+
+if [ -n "$DOCKER_APP_FAIL" ] ; then
+  echo "FAILING the fake docker app command"
+  exit 1
+fi
 
 if [ "$1" = "app" ] ; then
   echo "DOCKER-APP RENDER OUTPUT"
diff --git a/src/libaktualizr/package_manager/dockerappmanager.cc b/src/libaktualizr/package_manager/dockerappmanager.cc
index f30e22cd..2ae8a1ff 100644
--- a/src/libaktualizr/package_manager/dockerappmanager.cc
+++ b/src/libaktualizr/package_manager/dockerappmanager.cc
@@ -112,10 +112,13 @@ data::InstallationResult DockerAppManager::install(const Uptane::Target &target)
 }
 
 TargetStatus DockerAppManager::verifyTarget(const Uptane::Target &target) const {
+  TargetStatus status;
   if (target.IsOstree()) {
-    return OstreeManager::verifyTarget(target);
+    status = OstreeManager::verifyTarget(target);
+    if (status != TargetStatus::kGood) {
+      return status;
+    }
   }
-
   auto cb = [this](const std::string &app, const Uptane::Target &app_target) {
     LOG_INFO << "Verifying " << app << " -> " << app_target;
     std::stringstream ss;
diff --git a/src/libaktualizr/package_manager/dockerappmanager_test.cc b/src/libaktualizr/package_manager/dockerappmanager_test.cc
index 0e89ead5..44406217 100644
--- a/src/libaktualizr/package_manager/dockerappmanager_test.cc
+++ b/src/libaktualizr/package_manager/dockerappmanager_test.cc
@@ -100,6 +100,9 @@ TEST(DockerAppManager, DockerApp_Fetch) {
   client->package_manager_->install(target);
   std::string content = Utils::readFile(config.pacman.docker_apps_root / "app1/docker-compose.yml");
   ASSERT_EQ("DOCKER-APP RENDER OUTPUT\nfake contents of a docker app\n", content);
+
+  setenv("DOCKER_APP_FAIL", "1", 1);
+  ASSERT_EQ(TargetStatus::kInvalid, client->VerifyTarget(target));
 }
 
 #ifndef __NO_MAIN__
-- 
2.23.0

